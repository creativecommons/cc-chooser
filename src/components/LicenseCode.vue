<template>
    <p
        xmlns:dct="http://purl.org/dc/terms/"
        xmlns:cc="http://creativecommons.org/ns#"
        class="license-text"
    >
        <i18n
            path="license-use.richtext.full-text"
            tag="span"
        >
            <template v-slot:workTitle>
                <a
                    v-if="workUrl && isWeb"
                    :href="workUrl"
                    rel="cc:attributionURL"
                    :property="workTitle ? 'dct:title' : false"
                >
                    {{ workTitle }}
                </a>
                <span
                    v-else-if="workTitle"
                    property="dct:title"
                >
                    {{ workTitle }}
                </span>
                <span v-else>{{ $t('license-use.richtext.workTitle') }}</span>
            </template>
            <template v-slot:creator>
                <a
                    v-if="creatorName && creatorProfileUrl && isWeb"
                    :href="creatorProfileUrl"
                    rel="cc:attributionURL"
                    v-html="creatorSpan"
                />
                <span
                    v-else-if="creatorName"
                    v-html="creatorSpan"
                />
            </template>
            <template v-slot:by>
                {{ $t(byString) }}
            </template>
            <template v-slot:licensed-text>
                <span>{{ $t('license-use.richtext.licensed-text') }}</span>
            </template>
            <template v-slot:license>
                <a
                    v-if="isWeb"
                    :href="licenseUrl('web')"
                    target="_blank"
                    rel="license noopener noreferrer"
                    style="display: inline-block;"
                >
                    {{ shortName }}
                </a>
                <span v-else>{{ shortName }}</span>
                <LicenseIcons
                    v-if="isWeb"
                    :url="licenseUrl('web')"
                    :icons-arr="iconsList"
                />
            </template>
        </i18n>
        <i18n
            v-if="!isWeb"
            path="license-use.richtext.print-instructions"
            tag="span"
        >
            <template v-slot:linkToLicenseDeed>
                {{ licenseUrl('print') }}
            </template>
        </i18n>
    </p>
</template>
<script>
import { mapGetters, mapState } from 'vuex'
import LicenseIcons from './LicenseIcons'

export default {
    name: 'LicenseCode',
    components: {
        LicenseIcons
    },
    props: {
        attributionType: {
            type: String,
            default: 'web'
        }
    },
    computed: {
        ...mapGetters(['shortName', 'licenseUrl', 'iconsList']),
        ...mapState(['attributionDetails']),
        byString() {
            return this.creatorName ? 'license-use.richtext.by' : ''
        },
        creatorSpan() {
            if (this.creatorName) {
                return `<span property="cc:attributionName">${this.creatorName}</span>`
            } else return ''
        },
        creatorName() {
            return this.attributionDetails.creatorName
        },
        creatorProfileUrl() {
            const { creatorProfileUrl } = this.attributionDetails
            if (creatorProfileUrl && !creatorProfileUrl.startsWith('http')) {
                return `http://${creatorProfileUrl}`
            }
            return this.attributionDetails.creatorProfileUrl
        },
        workTitle() {
            return this.attributionDetails.workTitle
                ? this.attributionDetails.workTitle
                : this.$t('license-use.richtext.workTitle')
        },
        workUrl() {
            const { workUrl } = this.attributionDetails
            if (workUrl && !workUrl.startsWith('http')) {
                return `http://${workUrl}`
            }
            return workUrl
        },
        isWeb() {
            return this.attributionType === 'web'
        }
    }
}
</script>

<style scoped>
.license-text .photo-license-icons {
    height: 1.4rem!important;
    margin-left: 3px;
}
</style>
