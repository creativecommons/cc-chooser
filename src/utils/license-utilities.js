/** @typedef {{BY?: boolean, NC?: boolean, ND?: boolean, SA?: boolean}} LicenseAttributes */

/** @typedef {('CC0 1.0'|'CC BY 4.0'|'CC BY-NC 4.0'|'CC BY-NC-ND 4.0'|'CC BY-NC-SA 4.0'|'CC BY-ND 4.0'|'CC BY-ND-SA 4.0')} ShortLicenseName
 */
import {LICENSES} from './licenses'
const CC0Attributes = LICENSES.CC0.ATTRIBUTES
const CCBYAttributes = LICENSES.CC_BY.ATTRIBUTES
const defaultAttributes = { BY: undefined, NC: undefined, ND: undefined, SA: undefined }

/**
 * Convert short license name to attributes
 * @param {ShortLicenseName} shortLicenseName - name of the license with version number
 * @returns {LicenseAttributes}}
 */
function shortToAttr(shortLicenseName) {
    const short = shortLicenseName
    if (short.includes('CC0')) {
        return { ...CC0Attributes }
    }
    const nc = short.includes('NC')
    const nd = short.includes('ND')
    const sa = short.includes('SA')
    return { ...CCBYAttributes, NC: nc, ND: nd, SA: sa }
}

/**
 * Convert license attributes object to short license name
 * @param {LicenseAttributes} attr
 * @returns {ShortLicenseName}
 */
function attrToShort(attr) {
    if (attr.BY === undefined) return undefined
    if (!attr.BY) { return 'CC0 1.0' }
    let base = 'CC BY'
    if (attr.NC) { base += '-NC' }
    if (!attr.ND && attr.SA) {
        base += '-SA'
    } else if (attr.ND) {
        base += '-ND'
    }
    base += ' 4.0'
    return base
}

/**
 * Convert license attributes object to full license name
 * @param {{LicenseAttributes}} attr
 * @returns {string|undefined}
 */
function attrToFull(attr) {
    if (attr.BY === undefined) { return undefined }
    if (!attr.BY) { return 'CC0 1.0 Universal' }
    let base = 'Attribution'
    if (attr.NC) { base += '-NonCommercial' }
    if (!attr.ND && attr.SA) {
        base += '-ShareAlike'
    } else if (attr.ND) {
        base += '-NoDerivatives'
    }
    base += ' 4.0 International'
    return base
}

/**
 * Returns url to license from short license name with version number (eg. 'CC BY 4.0')
 * @param attr {LicenseAttributes} license attributes object
 * @param [mode] {'web'| 'print'} (?ref=chooser-v1, target and rel are added to the end of the link if mode is web)
 * @returns {string} url of the license information page
 */
function licenseUrl(attr, mode) {
    const linkRef = mode === 'web' ? '/?ref=chooser-v1' : ''
    if (attr.BY === false) {
        return `https://creativecommons.org/publicdomain/zero/1.0${linkRef}`
    }
    let short = attrToShort(attr).toLowerCase().slice(3)
    short = short.slice(0, short.length - 4)
    return `https://creativecommons.org/licenses/${short}/4.0${linkRef}`
}

/**
 * Convert short license name to licence slug ('CC BY 4.0' -> 'cc-by')
 * @param {ShortLicenseName} shortLicenseName
 * @returns {string}
 */
function licenseSlug(shortLicenseName) {
    return shortLicenseName
        .toLowerCase()
        .replace(' ', '-')
        .slice(0, shortLicenseName.length - 4)
}

/**
 * Convert license attributes object to an array of icon names
 * @param {LicenseAttributes} licenseAttributes
 * @returns {string[]|[]} Array with slugified names of icons, eg. ['cc', 'by']
 */
function licenseIconsArr(licenseAttributes) {
    if (!licenseAttributes.BY) {
        return ['zero']
    }
    const iconsArray = []
    for (const key in licenseAttributes) {
        if (licenseAttributes[key]) {
            iconsArray.push(key.toLowerCase())
        }
    }
    return iconsArray
}

function updateVisibleEnabledStatus(stepStatusData) {
    let visible
    let enabled
    let stepsDisabledDue = ''
    if (stepStatusData.FS) {
        // User will select from the dropdown
        if (stepStatusData.BY === false) {
            // User selected a license from the dropdown a CC0 license
            // First step, dropdown and attribution details should be visible and enabled
            visible = ['FS', 'DD', 'CW', 'AD']
            enabled = ['FS', 'DD', 'CW', 'AD']
            stepsDisabledDue = 'CC0'
        } else {
            // User hasn't selected anything yet, or selected a BY license
            // First step, dropdown and attribution details should be visible and enabled
            visible = ['FS', 'DD', 'AD']
            enabled = ['FS', 'DD', 'AD']
        }
    } else {
        // User uses the stepper for license selection
        if (stepStatusData.BY === false) {
            // User selects a CC0 license
            visible = ['FS', 'BY', 'NC', 'ND', 'SA', 'CW', 'AD']
            enabled = ['FS', 'BY', 'CW', 'AD']
            stepsDisabledDue = 'CC0'
        } else if (stepStatusData.ND) {
            // User selects an ND license: SA step is disabled
            visible = ['FS', 'BY', 'NC', 'ND', 'SA', 'AD']
            enabled = ['FS', 'BY', 'NC', 'ND', 'AD']
            stepsDisabledDue = 'CC0'
        } else {
            // User selects a non-ND BY license from the stepper
            visible = ['FS', 'BY', 'NC', 'ND', 'SA', 'AD']
            enabled = ['FS', 'BY', 'NC', 'ND', 'SA', 'AD']
        }
    }
    return { visible, enabled, stepsDisabledDue }
}


const CC_NAMESPACE = {
    NAME: 'xmlns:cc',
    URI: 'http://creativecommons.org/ns#'
}
const DCT_NAMESPACE = {
    NAME: 'xmlns:dct',
    URI: 'http://purl.org/dc/terms/'
}
const ICON_STYLE = 'height:22px!important;margin-left:3px;vertical-align:text-bottom;'
const ICON_BASE_URL = 'https://mirrors.creativecommons.org/presskit/icons/'

/**
 * Generate data for use in attribution HTML through i18n
 * @param attributionDetails
 * @param {ShortLicenseName} shortLicenseName
 * @returns {{creator: string, workTitle: string, licenseLink: string, htmlString: string}}
 */
function generateHTML(attributionDetails, shortLicenseName) {
    const dataForHtmlGeneration = {
        htmlString: '',
        creator: '',
        workTitle: '',
        licenseLink: ''
    }
    const { creatorName, creatorProfileUrl, workTitle, workUrl } = attributionDetails
    dataForHtmlGeneration.htmlString =
        `<p ${DCT_NAMESPACE.NAME}="${DCT_NAMESPACE.URI}"` +
        ` ${CC_NAMESPACE.NAME}="${CC_NAMESPACE.URI}"` +
        ' class="license-text">'
    const iconStyle = `style="${ICON_STYLE}"`
    const assetPathRef = '?ref=chooser-v1'
    let licenseIcons = `<img ${iconStyle} src="${ICON_BASE_URL}/cc.svg${assetPathRef}" />`
    const nameForSlug = shortLicenseName.includes('CC0') ? 'CC CC0 1.0' : shortLicenseName
    licenseIcons += nameForSlug
        .slice(3, nameForSlug.length - 4)
        .split('-')
        .map(attr => `<img ${iconStyle} src="${ICON_BASE_URL}/${attr.toLowerCase()}.svg${assetPathRef}" />`
        ).join('')
    const licenseHref = licenseUrl(shortToAttr(shortLicenseName))
    dataForHtmlGeneration.licenseLink =
        `<a rel="license" href="${licenseHref}" target="_blank"
        rel="license noopener noreferrer" style="display:inline-block;">
        ${shortLicenseName}${licenseIcons}</a>`

    if (creatorName) {
        if (creatorProfileUrl) {
            const absoluteUrl = creatorProfileUrl.startsWith('http') ? creatorProfileUrl : `http://${creatorProfileUrl}`
            dataForHtmlGeneration.creator =
                `<a rel="cc:attributionURL dct:creator" property="cc:attributionName" href="${absoluteUrl}">${creatorName}</a>`
        } else {
            dataForHtmlGeneration.creator = `<span property="cc:attributionName">${creatorName}</span>`
        }
    }
    if (workTitle) {
        if (workUrl) {
            const absoluteUrl = workUrl.startsWith('http') ? workUrl : `http://${workUrl}`
            dataForHtmlGeneration.workTitle = `<a rel="cc:attributionURL" property="dct:title" href="${absoluteUrl}">${workTitle}</a>`
        } else {
            dataForHtmlGeneration.workTitle = `<span rel="dct:title">${workTitle}</span>`
        }
    }
    return dataForHtmlGeneration
}

export {
    defaultAttributes, CC0Attributes, CCBYAttributes, shortToAttr, attrToShort,
    attrToFull, licenseUrl, licenseSlug, licenseIconsArr, generateHTML, updateVisibleEnabledStatus,
    CC_NAMESPACE, DCT_NAMESPACE, LICENSES, ICON_STYLE, ICON_BASE_URL
}
