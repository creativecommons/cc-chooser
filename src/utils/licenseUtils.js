// eslint brace-style: 0

function shortToFull(shortLicenseName) {
    var base = 'Attribution'
    var short = shortLicenseName.toLowerCase()
    if (short.includes('nc')) { base += '-NonCommercial' }
    if (short.includes('sa')) { base += '-ShareAlike' } else if (short.includes('nd')) { base += '-NoDerivatives' }
    // eslint-disable-next-line
    return base += ' 4.0'
}

function fullToShort(fullLicenseName) {
    var base = 'CC BY'
    var full = fullLicenseName
    if (full.includes('noncommercial')) { base += '-NC' }
    if (full.includes('sharealike')) { base += '-SA' } else if (full.includes('noderivatives')) { base += '-ND' }
    // eslint-disable-next-line
    return base += ' 4.0'
}

export default { shortToFull, fullToShort }
