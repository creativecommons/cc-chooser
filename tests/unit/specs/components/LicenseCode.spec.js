import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import { mount, createLocalVue } from '@vue/test-utils'
import LicenseCode from '@/components/LicenseCode.vue'
import createStore from '@/store'
import { CCBYAttributes } from '@/utils/license-utilities'

const TEST_DATA = {
    creatorName: 'Jane Doe',
    creatorProfileUrl: 'www.author.com',
    workTitle: 'My work',
    workUrl: 'www.author.com/picture.jpg'
}

describe('LicenseCode.vue', () => {
    let wrapper
    let state
    let localVue

    // Vue i18n is looking for locale key in messages,
    // i.e. $t('app') becomes 'messages.<en>.app'
    const messages = {}
    messages.en = require('@/locales/en.json')
    beforeEach(() => {
        localVue = createLocalVue()
        localVue.use(VueI18n)
        localVue.use(Vuex)
        // License Code is only available after the User selects a license,
        // so we do not need to test blank license attributes
        state = { currentLicenseAttributes: CCBYAttributes }
        const store = createStore(state)
        const i18n = new VueI18n({
            locale: 'en',
            messages: messages
        })

        wrapper = mount(LicenseCode, {
            localVue,
            store,
            i18n
        })
    })

    it('it renders without any errors', () => {
        expect(wrapper.isVueInstance()).toBeTruthy()
    })

    it('has correct information when creator name and work title are provided', () => {
        wrapper.vm.$store.commit('setCreatorName', TEST_DATA.creatorName)
        wrapper.vm.$store.commit('setWorkTitle', TEST_DATA.workTitle)
        const titleElement = wrapper.find('[property="dct:title"]')
        expect(titleElement.text()).toEqual(TEST_DATA.workTitle)
        expect(titleElement.name()).toEqual('span')
        expect(Object.keys(titleElement.attributes()).length).toEqual(1)
        // const creatorElement = wrapper.find('[property="cc:attributionName"]')
        // expect(Object.keys(creatorElement.attributes()).length).toEqual(1)
        // expect(creatorElement.text()).toEqual(TEST_DATA.creatorName)
        // expect(creatorElement.name()).toEqual('span')
    })
    it('has correct information when only urls are provided', () => {
        wrapper.vm.$store.commit('setWorkTitle', '')
        wrapper.vm.$store.commit('setCreatorName', '')
        wrapper.vm.$store.commit('setCreatorProfileUrl', TEST_DATA.creatorProfileUrl)
        wrapper.vm.$store.commit('setWorkUrl', TEST_DATA.workUrl)
        const titleElement = wrapper.find('[rel="cc:attributionURL"]')
        expect(titleElement.text()).toEqual('This work')
        expect(titleElement.name()).toEqual('a')
        expect(titleElement.attributes().href).toEqual(TEST_DATA.workUrl)
        expect(Object.keys(titleElement.attributes()).length).toEqual(2)
        const creatorElement = wrapper.findAll('[property="cc:attributionName"]')
        expect(creatorElement.length).toEqual(0)
    })

    it('has correct information all data are provided', () => {
        wrapper.vm.$store.commit('setCreatorName', TEST_DATA.creatorName)
        wrapper.vm.$store.commit('setWorkTitle', TEST_DATA.workTitle)
        wrapper.vm.$store.commit('setCreatorProfileUrl', TEST_DATA.creatorProfileUrl)
        wrapper.vm.$store.commit('setWorkUrl', TEST_DATA.workUrl)
        const titleElement = wrapper.find('[property="dct:title"]')
        expect(titleElement.text()).toEqual(TEST_DATA.workTitle)
        expect(titleElement.attributes().href).toEqual(TEST_DATA.workUrl)
        expect(titleElement.attributes().rel).toEqual('cc:attributionURL')
        expect(titleElement.name()).toEqual('a')
        expect(Object.keys(titleElement.attributes()).length).toEqual(3)
        const creatorElement = wrapper.find('[property="cc:attributionName"]')
        // expect(Object.keys(creatorElement.attributes()).length).toEqual(3)
        // expect(creatorElement.text()).toEqual(TEST_DATA.creatorName)
        // expect(creatorElement.attributes().href).toEqual('http://' + TEST_DATA.creatorProfileUrl)
        // expect(creatorElement.attributes().rel).toEqual('cc:attributionURL')
        // expect(creatorElement.name()).toEqual('a')
    })
})
