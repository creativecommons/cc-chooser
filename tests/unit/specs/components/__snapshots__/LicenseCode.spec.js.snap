// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicenseCode.vue has correct information all data are provided 1`] = `
<p xmlns:dct="http://purl.org/dc/terms/" xmlns:cc="http://creativecommons.org/ns#" class="license-text"><span><a href="http://www.author.com/picture.jpg" rel="cc:attributionURL" property="dct:title">
                My work
            </a> 
            by
         <a href="http://www.author.com" rel="cc:attributionURL"><span property="cc:attributionName">Jane Doe</span></a><span> is licensed under </span><a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display: inline-block;">
    CC BY 4.0
  </a> <a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="noopener noreferrer" class="photo-license-icons"><img alt="CC icon" title="CC icon" src="../assets/license-icons/cc.svg" class="photo-license-icon"> <img alt="by icon" title="by icon" src="[object Object]" class="photo-license-icon"></a></span>
  <!---->
</p>
`;

exports[`LicenseCode.vue has correct information when creator name and work title are provided 1`] = `
<p xmlns:dct="http://purl.org/dc/terms/" xmlns:cc="http://creativecommons.org/ns#" class="license-text"><span><span property="dct:title">
                My work
            </span>
  by
  <span><span property="cc:attributionName">Jane Doe</span></span><span> is licensed under </span><a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display: inline-block;">
    CC BY 4.0
  </a> <a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="noopener noreferrer" class="photo-license-icons"><img alt="CC icon" title="CC icon" src="../assets/license-icons/cc.svg" class="photo-license-icon"> <img alt="by icon" title="by icon" src="[object Object]" class="photo-license-icon"></a></span>
  <!---->
</p>
`;

exports[`LicenseCode.vue has correct information when only urls are provided 1`] = `
<p xmlns:dct="http://purl.org/dc/terms/" xmlns:cc="http://creativecommons.org/ns#" class="license-text"><span><a href="http://www.author.com/picture.jpg" rel="cc:attributionURL" property="dct:title">
                My work
            </a> 
            by
         <a href="http://www.author.com" rel="cc:attributionURL"><span property="cc:attributionName">Jane Doe</span></a><span> is licensed under </span><a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display: inline-block;">
    CC BY 4.0
  </a> <a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="noopener noreferrer" class="photo-license-icons"><img alt="CC icon" title="CC icon" src="../assets/license-icons/cc.svg" class="photo-license-icon"> <img alt="by icon" title="by icon" src="[object Object]" class="photo-license-icon"></a></span>
  <!---->
</p>
`;

exports[`LicenseCode.vue has the expected UI 1`] = `
<p xmlns:dct="http://purl.org/dc/terms/" xmlns:cc="http://creativecommons.org/ns#" class="license-text"><span><span property="dct:title">
                This work
            </span>

  <span> is licensed under </span><a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display: inline-block;">
    CC BY 4.0
  </a> <a href="https://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="noopener noreferrer" class="photo-license-icons"><img alt="CC icon" title="CC icon" src="../assets/license-icons/cc.svg" class="photo-license-icon"> <img alt="by icon" title="by icon" src="[object Object]" class="photo-license-icon"></a></span>
  <!---->
</p>
`;
